### ISSUE TITLE
Component not rerendered when preserved

### ISSUE THREAD

**Issue by michael** [substance/substance#311]
Is it possible that components are not rerendered by default when they are preserved (and class did not change). I just had such a case during my panel refactorings. It seems it gets recycled as it should but actually a rerender should be triggered unless shouldRerender is defined and returns false.

Will follow up with an example from that updated codebase.

**Comment by michael** [comment]
You can reproduce using lens#panel-refactor. The relevant code is here:

https://github.com/substance/lens/blob/panel-refactor/LensWriter.js#L160-L161

**Comment by michael** [comment]
I now use props to propagate things for the ContextSection, but still no rerender happens. :(

Test with Lens#panel-refactor. Open References panel and click on the focus buttons. The app will react except for the panel itself, because ContextSection just does nothing.

**Comment by michael** [comment]
Fixed.

**Comment by obuchtala** [comment]
Analyzing the code in ContextSection, I found out that the taken approach is invalid.
During rendering, ContextSection derives props for its child, the panel component, dynamically from the parent's state. As this issue shows, the approach generates the expectation, that just by changing the parent state the panel would get rerendered. This is a violation of the contract, that **a component is rerendered only if props or state have changed**.
This contract is essential for an efficient reactive rendering implementation, as otherwise no assumptions can be made on when to bound recursive rendering.

The solution to this is to provide the ContextSection itself with appropriate information. If it depends on the parent's state, then the parent must provide its state as prop, or derive appriopriate props from its state.

Closing this as invalid.
